rspace/src/main/scala/coop/rchain/rspace/merger/StateChange.scala

```diff
@@ -0,0 +1,159 @@
+package coop.rchain.rspace.merger
+
+import cats.Monoid
+import cats.effect.concurrent.Ref
+import cats.effect.{Concurrent, Sync}
+import cats.syntax.all._
+import coop.rchain.rspace.channelStore.ChannelStore
+import coop.rchain.rspace.hashing.{Blake2b256Hash, StableHashProvider}
+import coop.rchain.rspace.history.HistoryReaderBinary
+import coop.rchain.rspace.merger.MergingLogic.{consumesAffected, producesAffected}
+import coop.rchain.rspace.serializers.ScodecSerialize.{JoinsB, WaitingContinuationB}
+import coop.rchain.rspace.syntax._
+import coop.rchain.shared.Serialize
+import coop.rchain.shared.syntax._
+import fs2.Stream
+import scodec.bits.ByteVector
+
+/** Waiting continuation (WK) with channels of consume that put this WK. This class is used to compute joins changes. */
+final case class ConsumeChange(body: ByteVector, channels: Seq[Blake2b256Hash])
+
+/** We need this index to compute joins changes base on consume changes. This is because joins store not hashes of
+  * channels but channels themselves. So when we have channels from consumes (which are hashes of channels) -
+  * we are not able to reconstruct full join ByteVector */
+final case class JoinIndex(body: ByteVector, channels: Seq[Blake2b256Hash])
+
+final case class StateChange(
+    datumChanges: Map[Blake2b256Hash, ChannelChange[ByteVector]],
+    kontChanges: Map[Blake2b256Hash, ChannelChange[ConsumeChange]],
+    joinsIndex: Set[JoinIndex]
+)
+
+object StateChange {
+
+  private def computeValueChangeOnChannel[F[_]: Sync, T](
+      valuePointer: Blake2b256Hash,
+      getStartValue: Blake2b256Hash => F[Seq[T]],
+      getEndValue: Blake2b256Hash => F[Seq[T]]
+  ): F[ChannelChange[T]] =
+    for {
+      startValue <- getStartValue(valuePointer)
+      endValue   <- getEndValue(valuePointer)
+      added      = endValue diff startValue
+      deleted    = startValue diff endValue
+    } yield ChannelChange(added.toVector, deleted.toVector)
+
+  private def recordValueChange[F[_]: Concurrent, T](
+      valuePointer: Blake2b256Hash,
+      startValueReader: Blake2b256Hash => F[Seq[T]],
+      endValueReader: Blake2b256Hash => F[Seq[T]],
+      accumulator: Ref[F, Map[Blake2b256Hash, ChannelChange[T]]]
+  ): Stream[F, Unit] =
+    fs2.Stream.eval(for {
+      r <- computeValueChangeOnChannel[F, T](
+            valuePointer,
+            h => startValueReader(h),
+            h => endValueReader(h)
+          )
+      _ <- accumulator.update(s => {
+            val curVal = s.getOrElse(valuePointer, ChannelChange.empty)
+            val newVal = curVal.copy(
+              added = curVal.added ++ r.added,
+              removed = curVal.removed ++ r.removed
+            )
+            s.updated(valuePointer, newVal)
+          })
+    } yield ())
```

Comment:
This function can return `F[Unit]`. Caller can easily create singleton stream.

Reply:
https://github.com/rchain/rchain/pull/3410/commits/e11ed628180c1950e81e0e6186a86994838a31f2